bookkeeper-server/src/main/java/org/apache/bookkeeper/client/RackawareEnsemblePlacementPolicyImpl.java

```diff
@@ -1253,28 +1298,31 @@ static void shuffleWithMask(DistributionSchedule.WriteSet writeSet,
         }
     }
 
+    // this method should be called in readlock scope of 'rwlock'
     @Override
     public boolean isEnsembleAdheringToPlacementPolicy(List<BookieSocketAddress> ensembleList, int writeQuorumSize,
             int ackQuorumSize) {
         int ensembleSize = ensembleList.size();
         int minNumRacksPerWriteQuorumForThisEnsemble = Math.min(writeQuorumSize, minNumRacksPerWriteQuorum);
-        HashSet<String> racksOrRegionsInQuorum = new HashSet<String>();
+        HashSet<String> racksInQuorum = new HashSet<String>();
         BookieSocketAddress bookie;
         for (int i = 0; i < ensembleList.size(); i++) {
-            racksOrRegionsInQuorum.clear();
+            racksInQuorum.clear();
             for (int j = 0; j < writeQuorumSize; j++) {
                 bookie = ensembleList.get((i + j) % ensembleSize);
                 try {
-                    racksOrRegionsInQuorum.add(knownBookies.get(bookie).getNetworkLocation());
+                    racksInQuorum.add(knownBookies.get(bookie).getNetworkLocation());
                 } catch (Exception e) {
                     /*
-                     * any issue/exception in analyzing whether ensemble is strictly adhering to
-                     * placement policy should be swallowed.
+                     * any issue/exception in analyzing whether ensemble is
+                     * strictly adhering to placement policy should be
+                     * swallowed.
                      */
                     LOG.warn("Received exception while trying to get network location of bookie: {}", bookie, e);
                 }
             }
-            if (racksOrRegionsInQuorum.size() < minNumRacksPerWriteQuorumForThisEnsemble) {
+            if ((racksInQuorum.size() < minNumRacksPerWriteQuorumForThisEnsemble)
```

Comment:
Don't you need to check for enforceMinNumRacksPerWriteQuorum for this if condition too? If this is implicit why not the defaultRack is not implicit?

Reply:
didn't get you, here i'm making sure that number of racks in quorum is not less than minNumRacksPerWriteQuorumForThisEnsemble and if enforceMinNumRacksPerWriteQuorum is enabled then no bookie is from defaultRack.

Reply:
If enforceMinNumRacksPerWriteQuorum is not enabled, (racksInQuorum.size() < minNumRacksPerWriteQuorumForThisEnsemble) is not an issue right? So I am thinking the check should be enforceMinNumRacksPerWriteQuorum && ((racksInQuorum.size() < minNumRacksPerWriteQuorumForThisEnsemble) || (enforceMinNumRacksPerWriteQuorum && racksInQuorum.contains(getDefaultRack()))

Reply:
@jvrao If enforceMinNumRacksPerWriteQuorum is not enabled, (racksInQuorum.size() < minNumRacksPerWriteQuorumForThisEnsemble is **an** issue . for writequorum to adhere to placement policy it should have bookies from atleast minNumRacksPerWriteQuorumForThisEnsemble. if enforceMinNumRacksPerWriteQuorum is enabled then on top of the existing condition, no bookie should be from default rack.